genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobTerminationServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.rpc.grpc.services.impl.v4;
+
+import com.google.common.collect.Maps;
+import com.netflix.genie.common.exceptions.GenieException;
+import com.netflix.genie.common.exceptions.GenieServerException;
+import com.netflix.genie.proto.JobKillListenerRequest;
+import com.netflix.genie.proto.JobKillListenerResponse;
+import com.netflix.genie.proto.JobTerminationServiceGrpc;
+import com.netflix.genie.web.rpc.grpc.interceptors.SimpleLoggingInterceptor;
+import com.netflix.genie.web.services.JobKillServiceV4;
+import com.netflix.genie.web.services.JobSearchService;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import net.devh.springboot.autoconfigure.grpc.server.GrpcService;
+
+import javax.validation.constraints.NotBlank;
+import java.util.Map;
+
+/**
+ * Extension of {@link JobTerminationServiceGrpc.JobTerminationServiceImplBase}
+ * to provide functionality for killing jobs running on agents.
+ * TODO Register with HeartBeatService to listen for agent stream to become inactive and clean it up.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@GrpcService(
+    value = JobTerminationServiceGrpc.class,
+    interceptors = {
+        SimpleLoggingInterceptor.class,
+    }
+)
+@Slf4j
+public class GRpcJobTerminationServiceImpl
+    extends JobTerminationServiceGrpc.JobTerminationServiceImplBase
+    implements JobKillServiceV4 {
+
+    private final Map<String, StreamObserver<JobKillListenerResponse>> activeStreams = Maps.newConcurrentMap();
+    private final JobSearchService jobSearchService;
```

Comment:
Why job search service if you're always looking up records by ID?
Did Tom suggest this? Seems strange to me.

Reply:
Its an already existing service. Only using it.